src/szk.rs

```diff
@@ -0,0 +1,444 @@
+    /// each measurement.
+    pub fn new_turboshake128(typ: T, algorithm_id: u32) -> Self {
+        Szk::new(typ, algorithm_id)
+    }
+}
+
+impl<T, P, const SEED_SIZE: usize> Szk<T, P, SEED_SIZE>
+where
+    T: Type,
+    P: Xof<SEED_SIZE>,
+{
+    /// Construct an instance of this sharedZK proof system with the underlying FLP.
+    pub fn new(typ: T, algorithm_id: u32) -> Self {
+        Self {
+            typ,
+            algorithm_id,
+            phantom: PhantomData,
+        }
+    }
+
+    fn adjust_domsep_tag(&self, usage: u16) -> [u8; 8] {
+        let mut adjusted_dst = [0u8; 8];
+        adjusted_dst[0] = MASTIC_VERSION;
+        adjusted_dst[1] = 0; // algorithm class
+        adjusted_dst[2..6].copy_from_slice(&(self.algorithm_id).to_be_bytes());
+        adjusted_dst[6..8].copy_from_slice(&usage.to_be_bytes());
+        adjusted_dst
+    }
+
+    /// Derive a vector of random field elements for consumption by the FLP prover.
+    fn derive_prove_rands(&self, prove_rand_seed: &Seed<SEED_SIZE>) -> Vec<T::Field> {
+        P::seed_stream(
+            prove_rand_seed,
+            &self.adjust_domsep_tag(DST_PROVE_RANDOMNESS),
+            &[],
+        )
+        .into_field_vec(self.typ.prove_rand_len())
+    }
+
+    fn derive_joint_rand_part(
+        &self,
+        aggregator_blind: &Seed<SEED_SIZE>,
+        measurement_share: &[T::Field],
+        nonce: &[u8; 16],
+    ) -> Result<Seed<SEED_SIZE>, SzkError> {
+        let mut xof = P::init(
+            aggregator_blind.as_ref(),
+            &self.adjust_domsep_tag(DST_JOINT_RAND_PART),
+        );
+        // Encode measurement_share (currently an array of field elements) into bytes
+        // and include it in the XOF input.
+        let mut bytes = vec![0u8; T::Field::ENCODED_SIZE];
+        for field_elem in measurement_share {
+            field_elem.encode(&mut bytes)?;
+            xof.update(&bytes)
+        }
+        xof.update(nonce);
+        Ok(xof.into_seed())
+    }
+
+    fn derive_joint_rand_seed(
+        &self,
+        leader_joint_rand_part: &Seed<SEED_SIZE>,
+        helper_joint_rand_part: &Seed<SEED_SIZE>,
+    ) -> Seed<SEED_SIZE> {
+        let mut xof = P::init(
+            &[0; SEED_SIZE],
+            &self.adjust_domsep_tag(DST_JOINT_RAND_SEED),
+        );
+        xof.update(&leader_joint_rand_part.0);
+        xof.update(&helper_joint_rand_part.0);
+        xof.into_seed()
+    }
+
+    fn derive_joint_rands(
+        &self,
+        leader_joint_rand_part: &Seed<SEED_SIZE>,
+        helper_joint_rand_part: &Seed<SEED_SIZE>,
+    ) -> (Seed<SEED_SIZE>, Vec<T::Field>) {
+        let joint_rand_seed =
+            self.derive_joint_rand_seed(leader_joint_rand_part, helper_joint_rand_part);
+        let joint_rands = P::seed_stream(
+            &joint_rand_seed,
+            &self.adjust_domsep_tag(DST_JOINT_RANDOMNESS),
+            &[0],
```

Comment:
As with proof randomness, I think we could use an empty binder string here too. It should work either way, as this is the only XOF call with this usage constant, but the empty string is easier to explain.
```suggestion
            &[],
```